eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php

```diff
@@ -1160,4 +1174,36 @@ public function bulkRemoveTranslation($languageId, $actions)
         ;
         $query->execute();
     }
+
+    /**
+     * Cleanup obsolete lang_mask entries for the given Location.
+     *
+     * @param int $locationId
+     * @param int[] $newLanguageIds
+     * @param bool $alwaysAvailable
+     */
+    public function cleanupTranslationsAfterPublish($locationId, $newLanguageIds, $alwaysAvailable)
+    {
+        $connection = $this->dbHandler->getConnection();
+        /** @var \Doctrine\DBAL\Connection $connection */
+        $query = $connection->createQueryBuilder();
+        $query
+            ->select('id', 'lang_mask', 'parent', 'text_md5')
+            ->from($this->table)
+            ->where('action = :action')
+            ->setParameter('action', 'eznode:' . $locationId);
+
+        $statement = $query->execute();
+
+        // remove specific languages from a bit mask
+        foreach ($statement->fetchAll() as $row) {
+            $oldLanguageIds = $this->languageMaskGenerator->extractLanguageIdsFromMask((int) $row['lang_mask']);
+            foreach ($oldLanguageIds as $oldLanguageId) {
+                if (!in_array($oldLanguageId, $newLanguageIds)) {
+                    // note: it's impossible to determine single $newId, so setting it to refer to itself
+                    $this->handleRemovedTranslation($row['lang_mask'], $oldLanguageId, $row['parent'], $row['text_md5'], $row['id']);
```

Comment:
The purpose of the last argument, `$newId`, stored in `ezurlalias_ml`.`link` db table / column is to link archived (a.k.a. "historized") alias with the current one. However in this case, if we have e.g. three translations, one of them being removed, it's not possible to determine which translation archived alias should link to.

I'm setting here self reference to be FK friendly (once we have them).

Other ideas welcomed.